client/src/components/Map/Map.js

```diff
@@ -1,14 +1,27 @@
 import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
 import PropTypes from 'prop-types';
 import uniq from 'lodash/uniq';
+import { createSelector } from 'reselect';
+import cookies from 'browser-cookies';
 
 import SuperBlock from './components/SuperBlock';
 import Spacer from '../helpers/Spacer';
 
 import './map.css';
 import { ChallengeNode } from '../../redux/propTypes';
+import { toggleSuperBlock, toggleBlock, isInitializedSelector } from './redux';
+import {
+  currentChallengeUrlSelector,
+  currentChallengeIdSelector
+} from '../../redux';
+import { getBlocksFromChallengeUrl } from '../../utils';
+import { blockNameify } from '../../../utils/blockNameify';
 
 const propTypes = {
+  currentChallengeId: PropTypes.string,
```

Comment:
I do not see the use of `currentChallengeId`, but I think that it should be used.

Reply:
That was used in the first version of this PR, so you could search for the challenge's URL.  It's definitely not necessary any more, though, since we'll be using local storage or just passing the `challengeUrl` directly. 

I'll just take it out.  Unless I'm missing something!

Reply:
We get `currentChallengeId` from a user data, it can be more actual than `challengeUrl`.

Reply:
If you mean more recent, I'm not sure that's true.  As far as I can see, user data doesn't get refreshed unless you reload the page.  Also, wasn't the idea to use local storage instead? After all, updating based on the server response can cause the page to change while the user tries to click on something.

Reply:
I think we are not on the same page here. I just went through the conversation again.

The `currentChallengeId` is the last challenge that the user has submitted successfully and is on the user records. It is meant to be updated only when user clicks the submit button (and passes the challenge correctly).

We need not update that key anytime else. That is by design.

The navigation state can be anything else and should be local to the session. We should not need to care about syncing sessions (rare edge case).

This does bring some ambiguity on how to gauge user intent on the each time the map loads in the session. I think I like the idea of simply checking the localStorage / cookie and expanding the block if it is present and if not opening up the last challenge they submitted based on currentChallengeId, and finally falling back to the default of the first block.

Reply:
Apologies about the confusion. In all honesty, I hope we should have named the key on user as `currentlySolvedChallengeId` !! But DB migration is a huge nightmare.

Reply:
You are not correct here, `currentChallengeId` is updated for registered users on a mount action of any challenge:

https://github.com/freeCodeCamp/freeCodeCamp/blob/fcfbe0314f4999ea3d57e5aa7309044c6d1eeb3e/client/src/templates/Challenges/redux/current-challenge-saga.js#L53

Reply:
No I agree, that is updated every time. 

But I don't think that is intended the design. What you have found is actually a bug. 

The intent I guess was to store the successfully submitted challenge as their current challenge. But then it got 're-used' to store the navigation as well 🤷‍♂ !

So now we already have `completedChallengeId` and `currentChallengeId`

Reply:
How about `lastViewedChallengeId`?  Something completely different, so we never mix them up.

Reply:
Actually, if we're using it for navigation, we'd use `lastViewedChallengeUrl`, but that general idea.

Reply:
Well, I would not want to rename or add a key to the User Object. DB migration is a nightmare. If we are able to achieve the intent with acceptable fallbacks, then that would be ideal.

Reply:
I don't see, why do we need `completedChallengeId`, we always can get it from `completedChallenges`? `currentChallengeId` is where we were the last time.

Reply:
Sure, I meant the store key though, i.e. what I'm calling `currentChallengeUrl`, rather than anything server side.

Just to clear up a couple of things.  

Firstly, should this PR only change what happens when the user reloads the page?  I.e. if they just navigate around, complete challenges or whatever, the expanded state stays as they left it. 

Secondly, should we go back to having the server add `currentChallengeUrl` to `sessionMeta`?  Or is it enough to use the store, so that if they reload a challenge learn will expand that block?

The latest commit only re-initialises the expanded state on page reload and does not rely on the server.  Instead it puts the url in the store and uses that.  So, if they reload a challenge, then go back to the curriculum, that challenge's block will be expanded (assuming no local storage/cookie).

The fallback goes cookie (though I'll change that to local storage) -> store value -> default first block.

Reply:
As I see it:
- on challenge mount, additionally, put `currentChallengeId` in the store and the local storage;
- on the Map, take `currentChallengeId` from the store and if it is not set then from local storage;
- find by `currentChallengeId` SuperBlock and Block to expand;
- and maybe scroll to a current challenge.

Reply:
I think we can avoid that one trip to the server. Let's add a commit to use localStorage instead of cookies, see how this behaves (on .dev) and then evaluate our decision.